flax/errors.py

```diff
@@ -59,6 +59,37 @@ def __init__(self, message):
     super().__init__(error_msg)
 
 
+#################################################
+# lazy_init.py errors                           #
+#################################################
+
+
+class LazyInitError(FlaxError):
+  """Lazy Init function has uncomputable return values.
+
+  This happens when passing an argument to lazy_init with ``jax.ShapeDtypeStruct``
+  that affects the initialized variables.
+  Make sure the init function only uses the shape and dtype or pass an
+  actual JAX array if this is impossible.
+
+  Example::
+
+    class Foo(nn.Module):
+      @compact
+      def __call__(self, x):
+        # This parameter depends on the input x
+        # this causes an error when using lazy_init.
+        k = self.param("kernel", lambda _: x)
+        return x * k
+    Foo().lazy_init(random.PRNGKey(0), jax.ShapeDtypeStruct((8, 4), jnp.float32))
```

Comment:
Maybe add that it does work if you use actual values (right?):

`Foo().lazy_init(random.PRNGKey(0), jnp.zeros((8, 4)))`

Reply:
You shouldn't do that though because you risk unnecessary compute again.